src/gateway.c

```diff
@@ -719,66 +737,66 @@ static void handle_exec_sql_next(struct gateway *g,
 				 bool done)
 {
 	tracef("handle exec sql next");
+	struct leader *l = g->leader;
+	PRE(l != NULL);
 	struct cursor *cursor = &req->cursor;
-	struct response_result response = { 0 };
-	sqlite3_stmt *stmt = NULL;
+	int schema = req->schema;
+	struct response_result response = {};
+	sqlite3_stmt *stmt;
+	const char *sql;
 	const char *tail;
 	int tuple_format;
-	uint64_t req_id;
 	int rv;
 
-	if (req->sql == NULL || strcmp(req->sql, "") == 0) {
-		goto success;
-	}
-
-	/* stmt will be set to NULL by sqlite when an error occurs. */
-	assert(g->leader != NULL);
-	rv = sqlite3_prepare_v2(g->leader->conn, req->sql, -1, &stmt, &tail);
-	if (rv != SQLITE_OK) {
-		tracef("exec sql prepare failed %d", rv);
-		failure(req, rv, sqlite3_errmsg(g->leader->conn));
-		goto done;
-	}
-
-	if (stmt == NULL) {
-		goto success;
-	}
-
-	if (!done) {
-		switch (req->schema) {
-			case DQLITE_REQUEST_PARAMS_SCHEMA_V0:
-				tuple_format = TUPLE__PARAMS;
-				break;
-			case DQLITE_REQUEST_PARAMS_SCHEMA_V1:
-				tuple_format = TUPLE__PARAMS32;
-				break;
-			default:
-				/* Should have been caught by handle_exec_sql */
-				assert(0);
+	tuple_format = schema == DQLITE_REQUEST_PARAMS_SCHEMA_V0 ?
+				 TUPLE__PARAMS :
+		       schema == DQLITE_REQUEST_PARAMS_SCHEMA_V1 ?
+				 TUPLE__PARAMS32 :
+		       (POST(false && "impossible"), 0);
+
+	for (;;) {
+		stmt = NULL;
+		sql = req->sql;
+		if (sql == NULL || strcmp(sql, "") == 0) {
+			goto success;
 		}
-		rv = bind__params(stmt, cursor, tuple_format);
+		/* stmt will be set to NULL by sqlite when an error occurs. */
+		rv = sqlite3_prepare_v2(l->conn, sql, -1, &stmt, &tail);
 		if (rv != SQLITE_OK) {
-			failure(req, rv, "bind parameters");
-			goto done_after_prepare;
+			tracef("exec sql prepare failed %d", rv);
+			failure(req, rv, sqlite3_errmsg(l->conn));
+			goto done;
+		}
+		if (stmt == NULL) {
+			goto success;
```

Comment:
I don't get the combination with `/* stmt will be set to NULL by sqlite when an error occurs. */`.

Reply:
It's not a great comment, if an error occurs then `rv != SQLITE_OK` and also `stmt` is set to `NULL`. If `rv == SQLITE_OK` and also `stmt == NULL` it means there was nothing to prepare in the given string, so we've executed all statements successfully. I'll fix the comment